include/seqan3/core/detail/strong_type.hpp

```diff
@@ -0,0 +1,546 @@
+ * the compiler would emit an error message, since the `error` type is not convertible to the `window_size` type and
+ * vice versa.
+ *
+ * ### Adding Skills
+ *
+ * In most cases it is sufficient to protect the user interface from misuse by using strong types. Within the
+ * implementation the underlying value can then be extracted using the `getter`-member functions of the
+ * seqan3::detail::strong_type class. However, there might be scenarios, where the strong type is exposed to the user
+ * to work with, but with a restricted set of operations that can be applied to the type.
+ * A familiar use case are the time typedefs of the [std::chrono](http://en.cppreference.com/w/cpp/header/chrono)
+ * library. It is convenient for the user to subtract two time values representing, for example seconds, to get the
+ * duration between two time points. But not all operations like modulo or multiplication are really useful for this.
+ * In order to add skills to the seqan3::detail::strong_type the typedef can be further specialized with
+ * operations from the seqan::detail::skill enum. The corresponding operators will be added if the underlying type
+ * supports these operations as well.
+ * For example, we could further specify our error type to support increment and decrement operations.
+ *
+ * ```cpp
+ * using error = detail::strong_type<unsigned, struct search_error_tag, detail::skill::increment | detail::skill::decrement>;
+ *
+ * error e{4};
+ * --e;
+ * ++e;
+ * ```
+ */
+template <typename unnamed_type, typename name_tag, skill skills = skill::none>
+class strong_type : public skills_base<strong_type<unnamed_type, name_tag, skills>, skills>
+{
+public:
+
+    /*!\name Constructor, destructor and assignment.
+     * \brief The standard functions are explicitly defaulted.
+     * \{
+     */
+    constexpr strong_type() = default;
+    constexpr strong_type(strong_type const &) = default;
+    constexpr strong_type(strong_type &&) = default;
+    constexpr strong_type & operator= (strong_type const &) = default;
+    constexpr strong_type & operator= (strong_type &&) = default;
+    ~strong_type() = default;
+
+    //!|brief Construction from underlying value type.
+    constexpr explicit strong_type(unnamed_type const & _value) : value(_value) /*noexcept(noexcept(value(_value)))*/
+    {}
+
+    //!|brief Move construction from underlying value type.
+    template<typename unnamed_type_ = unnamed_type>
+    //!\cond
+        requires !std::is_lvalue_reference_v<unnamed_type_>
+    //!\endcond
+    constexpr explicit strong_type(unnamed_type_ && _value) : value(std::move(_value))
+    {}
+    //!\}
+
+    /*!\name Accessor.
+    * \{
+    */
+    //!\brief Returns the underlying value.
+    constexpr unnamed_type & get() noexcept
+    {
+        return value;
+    }
+
+    //!\brief Returns the underlying value.
+    constexpr unnamed_type const & get() const noexcept
+    {
+        return value;
+    }
+    //!\}
```

Comment:
What do you think about adding `operator*` as accessor, too?

Reply:
Well, I am not so sure, as it suggests pointer semantics, while there is no pointer.
And `.get()` is not hard to type and clearly indicates the semantics of the type for the reader of the code.

Reply:
Well, things like an std::optional provide *, too, but I have no strong feelings about it...

Reply:
Mhmm, good point though. Also calling the getter `value` would fit the STL interface more

Reply:
On the other hand reference_wrapper also has `.get()`, which is more fitting I guess. So I stick to the current implementation.